doc/irtt.1
.\" Automatically generated by Pandoc 3.6.1
.\"
.TH "IRTT" "1" "July 17, 2026" "v0.9.2" "IRTT Manual"
.SH NAME
irtt \- Isochronous Round\-Trip Time
.SH SYNOPSIS
irtt \f[I]command\f[R] [\f[I]args\f[R]]
.PP
irtt help \f[I]command\f[R]
.SH DESCRIPTION
IRTT measures round\-trip time and other latency related metrics using
UDP packets sent on a fixed period, and produces both text and JSON
output.
.SH COMMANDS
.TP
\f[I]client\f[R]
runs the client
.TP
\f[I]server\f[R]
runs the server
.TP
\f[I]report\f[R]
emits results from a JSON file
.TP
\f[I]bench\f[R]
runs HMAC and fill benchmarks
.TP
\f[I]timer\f[R]
runs timer resolution test
.TP
\f[I]clock\f[R]
runs wall vs monotonic clock test
.TP
\f[I]sleep\f[R]
runs sleep accuracy test
.TP
\f[I]version\f[R]
shows the version
.SH EXAMPLES
After installing IRTT, start a server:
.IP
.EX
$ irtt server
IRTT server starting...
[ListenerStart] starting IPv6 listener on [::]:2112
[ListenerStart] starting IPv4 listener on 0.0.0.0:2112
.EE
.PP
While that\[cq]s running, run a client.
If no options are supplied, it will send a request once per second, like
ping.
Here we simulate a one minute G.711 VoIP conversation by using an
interval of 20ms and randomly filled payloads of 172 bytes:
.IP
.EX
$ irtt client \-i 20ms \-l 172 \-d 1m \-\-fill=rand \-\-sfill=rand \-q 192.168.100.10
[Connecting] connecting to 192.168.100.10
[Connected] connected to 192.168.100.10:2112

                         Min     Mean   Median      Max  Stddev
                         \-\-\-     \-\-\-\-   \-\-\-\-\-\-      \-\-\-  \-\-\-\-\-\-
                RTT  11.93ms  20.88ms   19.2ms  80.49ms  7.02ms
         send delay   4.99ms  12.21ms  10.83ms  50.45ms  5.73ms
      receive delay   6.38ms   8.66ms   7.86ms  69.11ms  2.89ms
                                                               
      IPDV (jitter)    782ns   4.53ms   3.39ms  64.66ms   4.2ms
          send IPDV    256ns   3.99ms   2.98ms  35.28ms  3.69ms
       receive IPDV    896ns   1.78ms    966µs  62.28ms  2.86ms
                                                               
     send call time   56.5µs   82.8µs           18.99ms   348µs
        timer error       0s   21.7µs           19.05ms   356µs
  server proc. time   23.9µs   26.9µs             141µs  11.2µs

                duration: 1m0s (wait 241.5ms)
   packets sent/received: 2996/2979 (0.57% loss)
 server packets received: 2980/2996 (0.53%/0.03% loss up/down)
     bytes sent/received: 515312/512388
       send/receive rate: 68.7 Kbps / 68.4 Kbps
           packet length: 172 bytes
             timer stats: 4/3000 (0.13%) missed, 0.11% error
.EE
.PP
In the results above, the client and server are located at two different
sites, around 50km from one another, each of which connects to the
Internet via point\-to\-point WiFi.
The client is 3km NLOS through trees located near its transmitter, which
is likely the reason for the higher upstream packet loss, mean send
delay and IPDV.
.SH BUGS
.IP \[bu] 2
Windows is unable to set DSCP values for IPv6.
.IP \[bu] 2
Windows is unable to set the source IP address, so
\f[CR]\-\-set\-src\-ip\f[R] may not be used on the server.
.SH LIMITATIONS
.RS
.PP
\[lq]It is the limitations of software that give it life.\[rq]
.IP
.EX
\-Me, justifying my limitations
.EE
.RE
.SS Isochronous (fixed period) send schedule
Currently, IRTT only sends packets on a fixed period, foregoing the
ability to simulate arbitrary traffic.
Accepting this limitation offers some benefits:
.IP \[bu] 2
It\[cq]s easy to implement
.IP \[bu] 2
It\[cq]s easy to calculate how many packets and how much data will be
sent in a given time
.IP \[bu] 2
It simplifies timer error compensation
.PP
Also, isochronous packets are commonly seen in VoIP, games and some
streaming media, so it already simulates an array of common types of
traffic.
.SS Fixed packet lengths for a given test
Packet lengths are fixed for the duration of the test.
While this may not be an accurate simulation of some types of traffic,
it means that IPDV measurements are accurate, where they wouldn\[cq]t be
in any other case.
.SS Stateful protocol
There are numerous benefits to stateless protocols, particularly for
developers and data centers, including simplified server design,
horizontal scalabity, and easily implemented zero\-downtime restarts.
However, in this case, a stateful protocol provides important benefits
to the user, including:
.IP \[bu] 2
Smaller packet sizes (a design goal) as context does not need to be
included in every request
.IP \[bu] 2
More accurate measurement of upstream vs downstream packet loss (this
gets worse in a stateless protocol as RTT approaches the test duration,
complicating interplanetary tests!)
.IP \[bu] 2
More accurate rate and test duration limiting on the server
.SS In\-memory results storage
Results for each round\-trip are stored in memory as the test is being
run.
Each result takes 77 bytes in memory (maybe 80 with alignment).
When in streaming mode, this isn\[cq]t a problem as a circular buffer is
used for round\-trip data.
But when in standard mode (needed for statistics and json output), this
limits the effective duration of the test, especially at very low send
intervals.
However, the advantages are:
.IP \[bu] 2
It\[cq]s easier to perform statistical analysis (like calculation of the
median) on fixed arrays than on running data values
.IP \[bu] 2
We don\[cq]t need to either send client timestamps to the server, or
maintain a local running window of sent packet info, because they\[cq]re
all in memory, no matter when server replies come back
.IP \[bu] 2
Not accessing the disk during the test to write test output prevents
inadvertently affecting the results
.IP \[bu] 2
It simplifies the API
.PP
Before streaming mode was implemented, a decision was made to use
32\-bit unsigned ints for sequence numbers, because in\-memory results
storage meant it would be impractical to use up the sequence number
space.
With streaming mode, it\[cq]s more likely that the sequence number space
could wrap around.
At a 10ms interval, the counter will wrap around in 497 days with
undefined results.
Keep that in mind if doing very long, low interval tests.
.SS 64\-bit received window
In order to determine per\-packet differentiation between upstream and
downstream loss, a 64\-bit \[lq]received window\[rq] may be returned
with each packet that contains the receipt status of the previous 64
packets.
This can be enabled using \f[CR]\-\-stats=window/both\f[R] with the irtt
client.
Its limited width and simple bitmap format lead to some caveats:
.IP \[bu] 2
Per\-packet differentiation is not available (for any intervening
packets) if greater than 64 packets are lost in succession.
These packets will be marked with the generic \f[CR]Lost\f[R].
.IP \[bu] 2
While any packet marked \f[CR]LostDown\f[R] is guaranteed to be marked
properly, there is no confirmation of receipt of the receive window from
the client to the server, so packets may sometimes be erroneously marked
\f[CR]LostUp\f[R], for example, if they arrive late to the server and
slide out of the received window before they can be confirmed to the
client, or if the received window is lost on its way to the client and
not amended by a later packet\[cq]s received window.
.PP
There are many ways that this simple approach could be improved, such as
by:
.IP \[bu] 2
Allowing a wider window
.IP \[bu] 2
Encoding receipt seqnos in a more intelligent way to allow a wider seqno
range
.IP \[bu] 2
Sending confirmation of window receipt from the client to the server and
re\-sending unreceived windows
.PP
However, the current strategy means that a good approximation of
per\-packet loss results can be obtained with only 8 additional bytes in
each packet.
It also requires very little computational time on the server, and
almost all computation on the client occurs during results generation,
after the test is complete.
It isn\[cq]t as accurate with late (out\-of\-order) upstream packets or
with long sequences of lost packets, but high loss or high numbers of
late packets typically indicate more severe network conditions that
should be corrected first anyway, perhaps before per\-packet results
matter.
Note that in case of very high packet loss, the \f[B]total\f[R] number
of packets received by the server but not returned to the client (which
can be obtained using \f[CR]\-\-stats=count\f[R]) will still be correct,
which will still provide an accurate \f[B]average\f[R] loss percentage
in each direction over the course of the test.
.SS Use of Go
IRTT is written in Go.
That carries with it:
.IP \[bu] 2
Non\-negligible system call overhead
.IP \[bu] 2
A larger executable size than with C
.IP \[bu] 2
Somewhat slower execution speed than C (although \c
.UR https://benchmarksgame.alioth.debian.org/u64q/compare.php?lang=go&lang2=gcc
not that much slower
.UE \c
)
.PP
However, Go also has characteristics that make it a good fit for this
application:
.IP \[bu] 2
Go\[cq]s target is network and server applications, with a focus on
simplicity, reliability and efficiency, which is appropriate for IRTT
.IP \[bu] 2
Memory footprint tends to be significantly lower than with some
interpreted languages
.IP \[bu] 2
It\[cq]s easy to support a broad array of hardware and OS combinations
.SH SEE ALSO
irtt\-client(1), irtt\-server(1)
.PP
\c
.UR https://github.com/heistp/irtt/
IRTT GitHub repository
.UE \c
.SH AUTHOR
Pete Heist \c
.MT pete@heistp.net
.ME \c
.PP
Many thanks to both Toke Høiland\-Jørgensen and Dave Täht from the \c
.UR https://www.bufferbloat.net/
Bufferbloat project
.UE \c
\ for their valuable advice.
Any problems in design or implementation are entirely my own.
.SH HISTORY
IRTT was originally written to improve the latency and packet loss
measurements for the excellent \c
.UR https://flent.org
Flent
.UE \c
\ tool.
Flent was developed by and for the \c
.UR https://www.bufferbloat.net/projects/
Bufferbloat
.UE \c
\ project, which aims to reduce \[lq]chaotic and laggy network
performance,\[rq] making this project valuable to anyone who values
their time and sanity while using the Internet.


CHANGES.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 0.9.2 - 2026-07-17

### Added

- Add support for streaming mode (-s), for infinite duration tests
- Add raw mode for fixed units output to stdout during test
- Use LookupIP for client hostname lookups and log on multiple addresses
- Use LookupIP for server bind addresses
- Move to Go 1.25, x/net 0.57 and x/sys 0.47
- Record a per-packet "late" field for out-of-order and add to the output json
- Add report command that reads and emits saved results
- Allow IPv6 addresses without brackets
- Set DSCP on all packets, including opening handshake and close packets
- Add `expected_packets_sent` field to JSON to support report command

### Changed

- Change default hybrid timer sleep factor from 0.95 to 0.9

### Removed

- Remove packet length from report output to support reading results from JSON
  (avoids unmarshaling large ClientConfig struct for one field)

## 0.9.1 - 2021-05-18

### Added

- Improve Windows time support by using `QueryPerformanceCounter` and
  `GetSystemTimePreciseAsFileTime`. Time precision on Windows 10 can now
  reach 100ns.
- Add syslog support for the server (`--syslog` flag).
- Add a [SmokePing](https://oss.oetiker.ch/smokeping/) probe, available from
  SmokePing 2.7.2. It's possible to copy the
  [code](https://github.com/oetiker/SmokePing/blob/master/lib/Smokeping/probes/IRTT.pm)
  into older versions.
- Update OM2P build to use the new
  [MIPS softfloat support](https://github.com/golang/go/issues/18162) in Go 1.10.

### Changed

- Re-write git history with new email address.
- Re-licensed to GPLv2.

### Removed

- Stop disabling GC on client and remove `--gc` flag from server as disabling GC
  no longer offers a demonstrable performance benefit. See
  [The Journey of Go's Garbage Collector](https://blog.golang.org/ismmkeynote).

### Fixed

- Fix potential client race at startup.
- Fix issue on platforms with timer resolution not accurate enough
  to measure server processing time (closes #13).
- Change handled interrupt signals to only `os.Interrupt` and `syscall.SIGTERM`
  to fix Plan 9 build.
- Various documentation fixes.

## 0.9.0 - 2018-02-11

### Added

- Server fills are now supported, and may be restricted on the server. See
  `--sfill` for the client and `--allow-fills` on the server. As an example, one
  can do `irtt client --fill=rand --sfill=rand -l 172 server` for random
  payloads in both directions. The server default is `--allow-fills=rand` so
  that arbitrary data cannot be relayed between two hosts. `server_fill` now
  appears under `params` in the JSON.
- Version information has been added to the JSON output.

### Changed

- Due to adoption of the [pflag](https://github.com/ogier/pflag) package, all long
  options now start with -- and must use = with values (e.g. `--fill=rand`).
  After the subcommand, flags and arguments may now appear in any order.
- `irtt client` syntax changes:
  - `-rs` is renamed to `--stats`
  - `-strictparams` is removed and is now the default. `--loose` may be used
    instead to accept and use server restricted parameters, with a warning.
  - `-ts` is renamed to `--tstamp`
  - `-qq` is renamed to `-Q`
  - `-fillall` is removed and is now the default. `--fill-one` may be used as
    a small optimization, but should rarely be needed.
- `irtt server` syntax changes:
  - `-nodscp` is renamed to `--no-dscp`
  - `-setsrcip` is renamed to `--set-src-ip`
- The communication protocol has changed, so clients and servers must both be
  updated.
- The handshake now includes a protocol version, which may change independently
  of the release version, and must match exactly between the client and server
  or the client will refuse to connect.
- The default server minimum interval is now `10ms`.
- The default client duration has been changed from `1h` to `1m`.
- Some author info was changed in the commit history, so the rewritten history
  must be fetched in all forks and any changes rebased.

## 0.1.0 - 2017-10-15

### Added

- Initial, untagged development release.


irtt.go
package irtt

import "os"

const exitCodeSuccess = 0

const exitCodeRuntimeError = 1

const exitCodeBadCommandLine = 2

const exitCodeDoubleSignal = 3

const defaultRaw = false

const defaultQuiet = false

const defaultReallyQuiet = false

const defaultHMACKey = ""

type command struct {
	name  string
	desc  string
	run   func([]string)
	usage func()
}

var commands []command

func registerCommand(name, desc string, run func([]string), usage func()) {
	commands = append(commands, command{name, desc, run, usage})
}

func getCommand(name string) *command {
	for _, c := range commands {
		if c.name == name {
			return &c
		}
	}
	return nil
}

func init() {
	registerCommand("client", "runs the client", runClientCLI, clientUsage)
	registerCommand("server", "runs the server", runServerCLI, serverUsage)
	registerCommand("report", "emits results from a JSON file", runReport,
		reportUsage)
	registerCommand("bench", "runs HMAC and fill benchmarks", runBench, nil)
	registerCommand("timer", "runs timer resolution test", runTimer, nil)
	registerCommand("clock", "runs wall vs monotonic clock test", runClock, nil)
	registerCommand("sleep", "runs sleep accuracy test", runSleep, nil)
	registerCommand("version", "shows the version", runVersion, nil)
}

// RunCLI runs the command line interface with the given arguments (typically
// os.Args).
func RunCLI(args []string) {
	if len(args) < 2 {
		usageAndExit(cliUsage, exitCodeBadCommandLine)
	}

	unknownCommandAndExit := func(cmd string) {
		printf("Error: unknown command %s\n", cmd)
		usageAndExit(cliUsage, exitCodeBadCommandLine)
	}

	if args[1] == "help" {
		if len(args) < 3 {
			usageAndExit(cliUsage, exitCodeBadCommandLine)
		}
		c := getCommand(args[2])
		if c == nil {
			unknownCommandAndExit(args[2])
		}
		if c.usage == nil {
			printf("%s: %s", c.name, c.desc)
			os.Exit(exitCodeSuccess)
		}
		usageAndExit(c.usage, exitCodeSuccess)
	}

	c := getCommand(args[1])
	if c == nil {
		unknownCommandAndExit(args[1])
	}
	c.run(args[2:])
}

func cliUsage() {
	setTabWriter(0)
	printf("irtt: measures round-trip time with isochronous UDP packets")
	printf("")
	printf("Usage:")
	printf("")
	printf("\t\t\t\tirtt command [arguments]")
	printf("\t\t\t\tirtt help command")
	printf("")
	printf("Commands:")
	printf("")
	for _, c := range commands {
		printf("\t\t\t\t%s\t%s\t", c.name, c.desc)
	}
}

func usageAndExit(usage func(), exitCode int) {
	if exitCode != exitCodeSuccess {
		printTo = os.Stderr
	}
	usage()
	flush()
	os.Exit(exitCode)
}
